backend/store/proxy.go

```diff
@@ -0,0 +1,67 @@
+package store
+
+import (
+	"context"
+	"sync/atomic"
+	"unsafe"
+
+	corev2 "github.com/sensu/sensu-go/api/core/v2"
+	"github.com/sensu/sensu-go/types"
+)
+
+// EventStoreProxy is a mechanism for providing an EventStore with a replaceable
+// underlying implementation. It uses an atomic so that calls are not impeded by
+// mutex overhead.
+type EventStoreProxy struct {
+	impl unsafe.Pointer
+
+	// guard against the store being garbage collected,
+	// which would cause a crash if there were no more references
+	// to it and the impl was dereferenced.
+	gcGuard EventStore
+}
+
+func NewEventStoreProxy(s EventStore) *EventStoreProxy {
+	return &EventStoreProxy{
+		impl:    unsafe.Pointer(&s),
+		gcGuard: s,
+	}
+}
+
+func (e *EventStoreProxy) do() EventStore {
+	return *((*EventStore)(atomic.LoadPointer(&e.impl)))
```

Comment:
Do you have any literature that might help me understand what this means?

Reply:
https://golang.org/ref/spec#Package_unsafe

Basically, what's happening here is the pointer to the store is being stored as a raw pointer in the `EventStoreProxy`, and is being loaded with the atomic package so that when it's replaced, there is no danger of a race condition.

The expression you are asking about is saying: convert this raw unsafe pointer to a pointer to an `EventStore` interface, and then dereference it, yielding an `EventStore`.

Reply:
![giphy](https://user-images.githubusercontent.com/10634078/59059956-d7c2a980-8854-11e9-8a5b-0c817ed1cb4f.gif)